examples/control.ctt

```diff
@@ -0,0 +1,52 @@
+module control where
+
+import prelude
+
+fmap (A B: U) (F: U -> U): (A -> B) -> F A -> F B = undefined
+
+-- Recursion/Corecursion
+
+data fix (F: U -> U) = Fix (point: F (fix F))
+
+unfix (F: U -> U): fix F -> F (fix F) = split Fix f -> f
+embed (F: U -> U): F (fix F) -> fix F = \(x: F (fix F)) -> Fix x
+
+data freeF   (F: U -> U) (A B: U)  = ReturnF (a: A) | BindF (f: F B)
+data cofreeF (F: U -> U) (A B: U)  = CoBindF (a: A)         (f: F B)
+data free    (F: U -> U) (A: U)    = Free    (_: fix (freeF     F A))
+data cofree  (F: U -> U) (A: U)    = CoFree  (_: fix (cofreeF   F A))
+
+unfree   (A: U) (F: U -> U): free   F A -> fix (freeF   F A) = split Free   a -> a
+uncofree (A: U) (F: U -> U): cofree F A -> fix (cofreeF F A) = split CoFree a -> a
+
+-- data free (A: U) (F: U -> U) = pure (a: A) | bind (x: F (free A F)) -- non fix version
+
+-- Control Type-Classes
+
+pure (A: U) (F: U -> U): U
+    = (return: A -> F A)
+    * Unit
+
+functor (A B: U) (F: U -> U): U
+    = (fmap: (A -> B) -> F A -> F B)
+    * Unit
```

Comment:
Why don't you use this later? And why don't you have the rules as part of the structure?

Reply:
This is simple:
```
functor_ (A B: U) (F: U -> U): U
    = (fmap: (A B: U) -> (A -> B) -> F A -> F B)
    * (id: (x: F A) -> Path (F A) (fmap A A (idfun A) x) (idfun (F A) x))
    * (compose: (C: U) (f: B -> C) (g: A -> B) (x: F A) ->
      Path (F C) (fmap A C (o A B C f g) x)
                 ((o (F A) (F B) (F C) (fmap B C f) (fmap A B g)) x)) * Unit
```

Reply:
Why is there a Unit in your structure? It's not needed...

Also, note that you can write the equality in `F A -> F C` and not parametrize over `x : F A`. I don't know if it's easier to work with, but it could be ok because of function extensionality.

Reply:
The reason I'm using Unit terminator is because in `Sigma` — `(x: A) * B` I can't name last field. And for clarity in naming I just prolongate the structure to one more field.

Reply:
Just so you know, reasoning about these things might be quite tricky if you don't assume that you are working with sets (for which the equality/Path type is a proposition). The reason is that when you compare two `functor_` you will have to compare also the proofs in the `compose` fields...

Reply:
I see about the `Unit` thing...  Just so you know, it could maybe be a good idea to write accessor functions to get the fields, instead of always using a series of `.1` and `.2`. The problem with this though is that without implicit arguments these become quite verbose...

Reply:
Yeah, for modeling type classes in Harper/Kiselov style would be definitely great to have at least numeric accessors. But for me .1 .2 is ok. No problem here. I do love mini cores too.